Dockerfile

```diff
@@ -1,201 +1,98 @@
+        *) echo "unsupported architecture"; exit 1 ;; \
+    esac && \
+    curl -fsSL https://deb.nodesource.com/setup_18.x | bash - && \
+    apt-get install -y nodejs && \
+    npm cache clean --force
+
+# Set workdir and create localstack user
+WORKDIR /opt/code/localstack
+RUN useradd -ms /bin/bash localstack && \
+    mkdir -p /var/lib/localstack /tmp/localstack && \
+    chmod 777 /var/lib/localstack /tmp/localstack && \
+    touch /tmp/localstack/.marker
+
+# Install basic global tools
+RUN pip install --no-cache-dir --upgrade virtualenv
+
+# Install entrypoint script
+COPY bin/docker-entrypoint.sh /usr/local/bin/
+COPY bin/hosts /etc/hosts
+
+# Environment variables
+ENV USER=localstack LANG=C.UTF-8 PYTHONUNBUFFERED=1
 
 # Install the latest version of awslocal globally
-RUN --mount=type=cache,target=/root/.cache \
-    pip3 install --upgrade awscli awscli-local requests
+RUN pip install --no-cache-dir --upgrade awscli awscli-local requests
 
-
-#
-# builder: Stage which installs the dependencies of LocalStack Community
-#
+# Builder stage: Install dependencies for LocalStack Community
 FROM base AS builder
-ARG TARGETARCH
-
-# Install build dependencies to base
-RUN --mount=type=cache,target=/var/cache/apt \
-    apt-get update && \
-        # Install dependencies to add additional repos
-        # g++ is a workaround to fix the JPype1 compile error on ARM Linux "gcc: fatal error: cannot execute ‘cc1plus’"
-        apt-get install -y gcc g++
+RUN apt-get update && \
+    apt-get install -y --no-install-recommends gcc g++ && \
+    apt-get clean && rm -rf /var/lib/apt/lists/*
 
-# upgrade python build tools
-RUN --mount=type=cache,target=/root/.cache \
-    (virtualenv .venv && . .venv/bin/activate && pip3 install --upgrade pip wheel setuptools)
+# Upgrade Python build tools
+RUN virtualenv .venv && . .venv/bin/activate && pip install --upgrade pip wheel setuptools
 
-# add files necessary to install runtime dependencies
-ADD Makefile pyproject.toml requirements-runtime.txt ./
-# add the localstack start scripts (necessary for the installation of the runtime dependencies, i.e. `pip install -e .`)
-ADD bin/localstack bin/localstack.bat bin/localstack-supervisor bin/
+# Add necessary files and install runtime dependencies
+COPY Makefile pyproject.toml requirements-runtime.txt ./
+COPY bin/localstack bin/localstack.bat bin/localstack-supervisor bin/
+RUN . .venv/bin/activate && pip install -r requirements-runtime.txt
 
-# Install dependencies for running the LocalStack runtime
-RUN --mount=type=cache,target=/root/.cache\
-    . .venv/bin/activate && pip3 install -r requirements-runtime.txt
-
-
-#
-# final stage: Builds upon base stage and copies resources from builder stages
-#
+# Final stage: Build LocalStack
 FROM base
 COPY --from=builder /opt/code/localstack/.venv /opt/code/localstack/.venv
-# The build version is set in the docker-helper.sh script to be the output of setuptools_scm
-ARG LOCALSTACK_BUILD_VERSION
```

Comment:
ARG commands are only applicable in the context of a single image, so this cannot be removed here (because the arg is used with the `setuptools_scm` env var below).